internal/app.go

```diff
@@ -22,51 +31,115 @@ const (
+	grpcConn *grpc.GrpcConnection
+	commit   string
+	version  string
 }
 
 func NewApp(commit, version string) *App {
-	return &App{commit, version}
+	return &App{nil, version, commit}
 }
 
-func (a *App) Run(ctx context.Context) error {
+func (a *App) Run(ctx context.Context) (err error) {
+	a.handleSignal(ctx)
+
 	config.Init(a.version, a.commit)
 
 	config.RegisterRunner(func(_ *cobra.Command, _ []string) {
-		err := config.RegisterConfigFile()
+		err = config.RegisterConfigFile()
 		if err != nil {
 			slog.ErrorContext(ctx, "Failed to load configuration file", "error", err)
 			return
 		}
 
-		agentConfig, err := config.ResolveConfig()
-		if err != nil {
-			slog.ErrorContext(ctx, "Invalid config", "error", err)
+		agentConfig, configErr := config.ResolveConfig()
+		if configErr != nil {
+			slog.ErrorContext(ctx, "Invalid config", "error", configErr)
 			return
 		}
 
-		slogger := logger.New(*agentConfig.Log)
-		slog.SetDefault(slogger)
+		standardLogger := logger.New(*agentConfig.Log)
+		slog.SetDefault(standardLogger)
 
 		slog.InfoContext(ctx, "Starting NGINX Agent",
 			slog.String("version", a.version),
 			slog.String("commit", a.commit),
 		)
 
 		messagePipe := bus.NewMessagePipe(defaultMessagePipeChannelSize)
-		err = messagePipe.Register(defaultQueueSize, plugin.LoadPlugins(ctx, agentConfig))
-		if err != nil {
-			slog.ErrorContext(ctx, "Failed to register plugins", "error", err)
+		msgPipeErr := messagePipe.Register(defaultQueueSize, a.loadPlugins(agentConfig))
+		if msgPipeErr != nil {
+			slog.ErrorContext(ctx, "Failed to register plugins", "error", msgPipeErr)
 			return
 		}
 
 		messagePipe.Run(ctx)
 	})
-	err := config.Execute(ctx)
+	err = config.Execute(ctx)
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
+
+func (a *App) handleSignal(ctx context.Context) {
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGINT, syscall.SIGTERM)
+
+	go func() {
+		select {
+		case <-sigChan:
+			slog.DebugContext(ctx, "NGINX Agent App exiting")
+		case <-ctx.Done():
+			if a.grpcConn != nil {
+				connCloseErr := a.grpcConn.Close(ctx)
+				slog.ErrorContext(ctx, "Issue closing gRPC connection", "error", connCloseErr)
+			}
+		}
+	}()
+}
+
+func (a *App) loadPlugins(agentConfig *config.Config) []bus.Plugin {
+	plugins := make([]bus.Plugin, 0)
+
+	plugins = a.addResourcePlugin(plugins, agentConfig)
+
+	if a.isGrpcClientConfigured(agentConfig) {
+		commandPlugin := command.NewCommandPlugin(agentConfig, a.grpcConn)
+		plugins = append(plugins, commandPlugin)
+		filePlugin := file.NewFilePlugin(agentConfig, a.grpcConn)
+		plugins = append(plugins, filePlugin)
+	}
+
+	plugins = a.addCollector(agentConfig, plugins)
+	plugins = append(plugins, watcher.NewWatcher(agentConfig))
+
+	return plugins
+}
+
+func (a *App) addResourcePlugin(plugins []bus.Plugin, agentConfig *config.Config) []bus.Plugin {
+	resourcePlugin := resource.NewResource(agentConfig)
+	plugins = append(plugins, resourcePlugin)
+
+	return plugins
+}
```

Comment:
nitpick on consistency: Does this need to be in its own function, since we are doing this same "init plugin and append to slice" combo in `loadPlugins` for other plugins?